pkg/daemon/server/service/rater/helper_test.go

```diff
@@ -329,158 +289,45 @@ func TestCalculateRate(t *testing.T) {
 		tc4.Update(&PodReadCount{"pod2", map[string]float64{"partition2": 400.0}})
 		tc4.Update(&PodReadCount{"pod100", map[string]float64{"partition100": 200.0}})
 		q.Append(tc4)
-		qItems := q.Items()
 
 		// partition1 rate
+		// no enough data collected within lookback seconds, expect 0 rate
 		assert.Equal(t, 0.0, CalculateRate(q, 5, "partition1"))
+		// no enough data collected within lookback seconds, expect 0 rate
 		assert.Equal(t, 0.0, CalculateRate(q, 15, "partition1"))
+		// tc2 and tc3 are used to calculate the rate
 		assert.Equal(t, 5.0, CalculateRate(q, 25, "partition1"))
+		// tc1, 2 and 3 are used to calculate the rate
 		assert.Equal(t, 7.5, CalculateRate(q, 35, "partition1"))
+		// tc1, 2 and 3 are used to calculate the rate
 		assert.Equal(t, 7.5, CalculateRate(q, 100, "partition1"))
-		assert.Equal(t, -150.0, getDeltaBetweenTimestampedCounts(qItems[0], qItems[2], "partition1"))
 
 		// partition2 rate
 		assert.Equal(t, 0.0, CalculateRate(q, 5, "partition2"))
 		assert.Equal(t, 0.0, CalculateRate(q, 15, "partition2"))
 		assert.Equal(t, 10.0, CalculateRate(q, 25, "partition2"))
 		assert.Equal(t, 10.5, CalculateRate(q, 35, "partition2"))
 		assert.Equal(t, 10.5, CalculateRate(q, 100, "partition2"))
-		assert.Equal(t, 210.0, getDeltaBetweenTimestampedCounts(qItems[0], qItems[2], "partition2"))
 
 		// partition3 rate
 		assert.Equal(t, 0.0, CalculateRate(q, 5, "partition3"))
 		assert.Equal(t, 0.0, CalculateRate(q, 15, "partition3"))
 		assert.Equal(t, 20.0, CalculateRate(q, 25, "partition3"))
-		assert.Equal(t, 7.5, CalculateRate(q, 35, "partition3"))
-		assert.Equal(t, 7.5, CalculateRate(q, 100, "partition3"))
-		assert.Equal(t, 150.0, getDeltaBetweenTimestampedCounts(qItems[0], qItems[2], "partition3"))
+		assert.Equal(t, 10.0, CalculateRate(q, 35, "partition3"))
+		assert.Equal(t, 10.0, CalculateRate(q, 100, "partition3"))
 
 		// partition4 rate
 		assert.Equal(t, 0.0, CalculateRate(q, 5, "partition4"))
 		assert.Equal(t, 0.0, CalculateRate(q, 15, "partition4"))
 		assert.Equal(t, 10.0, CalculateRate(q, 25, "partition4"))
 		assert.Equal(t, 5.0, CalculateRate(q, 35, "partition4"))
 		assert.Equal(t, 5.0, CalculateRate(q, 100, "partition4"))
-		assert.Equal(t, 100.0, getDeltaBetweenTimestampedCounts(qItems[0], qItems[2], "partition4"))
 
 		// partition100 rate
 		assert.Equal(t, 0.0, CalculateRate(q, 5, "partition100"))
 		assert.Equal(t, 0.0, CalculateRate(q, 15, "partition100"))
 		assert.Equal(t, 0.0, CalculateRate(q, 25, "partition100"))
 		assert.Equal(t, 0.0, CalculateRate(q, 35, "partition100"))
 		assert.Equal(t, 0.0, CalculateRate(q, 100, "partition100"))
```

Comment:
I remember @veds-g added these unit tests, to cover the edge cases since we are using binary search to fetch the start index. Please don't remove unit tests, it's always good to have more unit tests. The purpose of unit tests is to ensure each unit of the code performs as expected, so that we get confidence when we run it as a whole. If these edge cases are covered in other tests, I am ok with deleting these.

Reply:
> If these edge cases are covered in other tests, I am ok with deleting these.

+1. I will see if I can cover these edge cases in other tests.

> Please don't remove unit tests, it's always good to have more unit tests.

I have a mixed feeling about this statement. As I mentioned in PR description, unit tests should focus on testing what instead of how. I would say it's always good to have more unit tests to test what the function does. When we have a well-written unit test set to cover all the functionalities, any performance optimization code changes should not require any changes to unit tests because it doesn't change **what**.

Using our code as example, changing how we find start index from linear search to binary search is an optimization which shouldn't require unit test update. If it introduces bugs that are not caught by exiting unit tests, then we were not doing a good job testing.

Reply:
I just checked through the test cases. I confirm that ALL test cases that I deleted (`TestCalculatePartitionDelta` and `TestFindStartIndex`) are already covered in `TestCalculateRate`.